File: python/task3.py
import time
import json
import sys
from pyspark import SparkContext


class explore:
    def __init__(
        self,
        review_filepath,
        business_filepath,
        output_filepath_question_a,
        output_filepath_question_b,
    ):
        self.sc = SparkContext("local[*]", "task3")
        self.sc.setLogLevel("WARN")
        self.review_filepath = review_filepath
        self.business_filepath = business_filepath
        self.output_filepath_question_a = output_filepath_question_a
        self.output_filepath_question_b = output_filepath_question_b

    def data_preprocess(self):
        review_data = self.sc.textFile(review_filepath).map(
            lambda line: (json.loads(line)["business_id"], (json.loads(line)["stars"]))
        )
        business_data = self.sc.textFile(business_filepath).map(
            lambda line: (json.loads(line)["business_id"], (json.loads(line)["city"]))
        )
        data_joined = review_data.join(business_data)
        city_rating_data = data_joined.map(lambda line: (line[1][1], line[1][0]))
        return city_rating_data

    def avg_city_rating_spark_sort(self, city_rating_data):
        avg_by_key = (
            city_rating_data.mapValues(lambda v: (v, 1))
            .reduceByKey(lambda a, b: (a[0] + b[0], a[1] + b[1]))
            .mapValues(lambda v: v[0] / v[1])
            .sortBy(lambda x: (-x[1], x[0]))
            .collectAsMap()
        )
        # print(avg_by_key)
        return avg_by_key

    def avg_city_rating_python_sort(self, city_rating_data):
        avg_by_key = (
            city_rating_data.mapValues(lambda v: (v, 1))
            .reduceByKey(lambda a, b: (a[0] + b[0], a[1] + b[1]))
            .mapValues(lambda v: v[0] / v[1])
            .collectAsMap()
        )
        avg_by_key = sorted(avg_by_key.items(), key=lambda x: (-int(x[1]), x[0]))
        # print(avg_by_key)
        return avg_by_key

    def write_results(self, results, m1_time, m2_time, reason):
        f = open(self.output_filepath_question_a, "w")
        f.write("city,star\n")
        for key, value in results.items():
            f.write("{},{}\n".format(key, value))
        f.close
        answer = {"m1": m1_time, "m2": m2_time, "reason": reason}
        with open(self.output_filepath_question_b, "w") as json_file:
            json.dump(answer, json_file)


if __name__ == "__main__":
    review_filepath = sys.argv[1]
    business_filepath = sys.argv[2]
    output_filepath_question_a = sys.argv[3]
    output_filepath_question_b = sys.argv[4]
    obj = explore(
        review_filepath,
        business_filepath,
        output_filepath_question_a,
        output_filepath_question_b,
    )
    start_time = time.time()
    city_rating_data = obj.data_preprocess()
    m_time = time.time() - start_time
    start_time = time.time()
    result1 = obj.avg_city_rating_python_sort(city_rating_data)
    m1_time = time.time() - start_time
    start_time = time.time()
    result2 = obj.avg_city_rating_spark_sort(city_rating_data)
    m2_time = time.time() - start_time
    if dict(result1) == result2:
        reason = ""
        obj.write_results(result2, m1_time + m_time, m2_time + m_time, reason)

File: scala/src/main/scala/task3.scala
import org.apache.log4j.{Level, Logger}
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization
import java.io._
import scala.collection.mutable._

object task3 {
  def main(args: Array[String]): Unit = {
    Logger.getLogger("org").setLevel(Level.ERROR)
    Logger.getLogger("akka").setLevel(Level.ERROR)

    val review_filepath = args(0)
    val business_filepath = args(1)
    val output_filepath_question_a = args(2)
    val output_filepath_question_b = args(3)

    //val review_filepath = "../data/review.json"
    //val business_filepath = "../data/business.json"
    //val output_filepath_question_a = "abc3_scala.txt"
    //val output_filepath_question_b = "xyz3_scala.json"

    val spark = new SparkConf().setAppName("task3").setMaster("local[*]")
    val sc = new SparkContext(spark)
    def extract_stars(line: String): Double = {
      val data = parse(line)
      val date = data \ "stars"
      val JDouble(res) = date
      res.toDouble
    }
    def extract_string(line: String, col_name : String): String = {
      val data = parse(line)
      val date = data \ col_name
      val JString(res) = date
      res
    }
    def reduce_func(first : (Int, Double), second: (Int, Double)):(Int, Double)={
      (first._1 + second._1, first._2 + second._2)
    }
    def map_func(x : (String, (Int, Double))): (Double, String)={
      (x._2._2/x._2._1, x._1)
    }
    val start_time_m = System.nanoTime
    val review_data = sc.textFile(path = review_filepath)
      .map(line => (extract_string(line, "business_id") , extract_stars(line)))
    val business_data = sc.textFile(path = business_filepath)
      .map(line => (extract_string(line, "business_id"), extract_string(line, "city")))
    val rdd_data = review_data.join(business_data)
    val city_rating_data = rdd_data.map(line => (line._2._2, (1, line._2._1)))
    //city_rating_data.foreach(println)
    val m_time = (System.nanoTime - start_time_m) / 1e9d

    val start_time_m1 = System.nanoTime
    val average_1 = city_rating_data.reduceByKey(reduce_func).map(x => map_func(x)).collect()
    val m1_data = average_1.sortBy(average_1 => (-average_1._1, average_1._2))
    val m1_data_top10 = m1_data.slice(0,10)
    val m1_time = (System.nanoTime - start_time_m1) / 1e9d
    //m1_data.foreach(println)

    val start_time_m2 = System.nanoTime
    def ftn_(line: (Double, List[String])): List[(String, Double)] = {
      val num = line._1
      val list_of_names = line._2
      var res: List[(String, Double)] = List()
      for (w <- 0 to list_of_names.size - 1) {
        res = res :+ ((list_of_names(w), num))
      }
      return res
    }
    val m2_data = city_rating_data
      .reduceByKey(reduce_func)
      .map(x => map_func(x))
      .takeOrdered(num=10)(Ordering[(Double, String)].on(x => (-x._1, x._2)))

    val m2_time = (System.nanoTime - start_time_m2) / 1e9d
    m2_data.foreach(println)
    //println("M1 Method: " + m1_time)
    //println("M2 Method: " + m2_time)
    val file = new File(output_filepath_question_a)
    val bw = new BufferedWriter(new FileWriter(file))
    bw.write("city, stars"+"\n")
    for (w <- 0 to m1_data.size-1){
      bw.write(m1_data(w)._2.toString +','+ m1_data(w)._1.toString + "\n")
    }
    bw.close()

    val reason = ""
    val result = LinkedHashMap(
      "m1" -> (m1_time+m_time),
      "m2" -> (m2_time+m_time),
      "reason" -> reason)

    implicit val formats = org.json4s.DefaultFormats
    val writer = new PrintWriter(new File(output_filepath_question_b))
    writer.write(Serialization.write(result))
    writer.close()
  }
}
